datasets/noaa/_images/run_csv_transform_kub_ghcnd_inventory/csv_transform.py

```diff
@@ -0,0 +1,174 @@
+    ftp_host: str,
+    ftp_dir: str,
+    ftp_filename: str,
+    source_file: pathlib.Path,
+    target_file: pathlib.Path,
+    target_gcs_bucket: str,
+    target_gcs_path: str,
+) -> None:
+
+    # source_url            STRING          -> The full url of the source file to transform
+    # ftp_host              STRING          -> The host IP of the ftp file (IP only)
+    # ftp_dir               STRING          -> The remote working directory that the FTP file resides in (directory only)
+    # ftp_filename          STRING          -> The name of the file to pull from the FTP site
+    # source_file           PATHLIB.PATH    -> The (local) path pertaining to the downloaded source file
+    # target_file           PATHLIB.PATH    -> The (local) target transformed file + filename
+    # target_gcs_bucket     STRING          -> The target GCS bucket to place the output (transformed) file
+    # target_gcs_path       STRING          -> The target GCS path ( within the GCS bucket ) to place the output (transformed) file
+
+    logging.info("GCHND Inventory process started")
+
+    logging.info("Creating 'files' folder")
+    pathlib.Path("./files").mkdir(parents=True, exist_ok=True)
+
+    download_file_ftp(ftp_host, ftp_dir, ftp_filename, source_file, source_url)
+
+    df_filedata = load_source_file(source_file)
+
+    logging.info(f"Transformation Process Starting.. {source_file}")
+
+    df = extract_columns(df_filedata)
+    df = reorder_headers(df)
+
+    logging.info(f"Transformation Process complete .. {source_file}")
+
+    save_to_new_file(df, file_path=str(target_file))
+
+    logging.info(
+        f"Uploading output file to.. gs://{target_gcs_bucket}/{target_gcs_path}"
+    )
+    upload_file_to_gcs(target_file, target_gcs_bucket, target_gcs_path)
+
+    logging.info("GCHND Inventory process completed")
+
+
+def load_source_file(source_file: str) -> pd.DataFrame:
+    logging.info(f"Opening file {source_file}")
+    df_filedata = pd.read_csv(source_file, sep="|", header=None, names=["textdata"])
+    return df_filedata
+
+
+def reorder_headers(df: pd.DataFrame) -> pd.DataFrame:
+    logging.info("Transform: Reordering headers..")
+
+    df = df[
+        [
+            "id",
+            "latitude",
+            "longitude",
+            "element",
+            "firstyear",
+            "lastyear",
+        ]
+    ]
+
+    return df
+
+
+def extract_columns(df_filedata: pd.DataFrame) -> pd.DataFrame:
+    # Example:
+    # ZZZ99999999  17.1167  -61.7833 TMAX 1949 1949
+    col_ranges = {
+        "id": slice(0, 11),  # LENGTH: 12  EXAMPLE ID:  123456789999
+        "latitude": slice(12, 20),  # LENGTH:  9  EXAMPLE LAT: -31.0123
+        "longitude": slice(21, 30),  # LENGTH: 10  EXAMPLE LON: -31.0123
+        "element": slice(31, 35),  # LENGTH:  5  EXAMPLE ELE: ZZZZ
+        "firstyear": slice(36, 40),  # LENGTH:  5  EXAMPLE FIRSTYEAR: 1901
+        "lastyear": slice(41, 45),  # LENGTH:  5  EXAMPLE LASTYEAR: 1902
```

Comment:
Hope you can align the comments. I've done it for you below:

```
        "id": slice(0, 11),          # LENGTH: 12  EXAMPLE ID:  123456789999
        "latitude": slice(12, 20),   # LENGTH:  9  EXAMPLE LAT: -31.0123
        "longitude": slice(21, 30),  # LENGTH: 10  EXAMPLE LON: -31.0123
        "element": slice(31, 35),    # LENGTH:  5  EXAMPLE ELE: ZZZZ
        "firstyear": slice(36, 40),  # LENGTH:  5  EXAMPLE FIRSTYEAR: 1901
        "lastyear": slice(41, 45),   # LENGTH:  5  EXAMPLE LASTYEAR: 1902
```

Reply:
Done